File: bae-avalonia/bae-avalonia.ViewTests/FieldOriginDotTests.cs
using System.Linq;
using Avalonia.Controls;
using Avalonia.Controls.Shapes;
using Avalonia.Headless.XUnit;
using Avalonia.LogicalTree;
using Bae.Desktop;
using uniffi.bae_bridge;
using Xunit;

namespace Bae.Desktop.ViewTests;

/// <summary>What the dot after a release field says, and when it says
/// nothing.</summary>
public sealed class FieldOriginDotTests
{
    /// <summary>A field core marked draws a dot, and a field it marked nothing
    /// for draws none. Which of them are marked is core's answer.</summary>
    [AvaloniaFact]
    public void ADotIsDrawnForTheMarkedFieldsAndForNoOthers()
    {
        var marked = ImportCandidateFixtures.FieldProvenance()
            .Where(entry => FieldOriginDot.For(entry) is not null)
            .Select(entry => entry.Field)
            .ToArray();

        Assert.Equal(
            new[]
            {
                BridgeCandidateEditField.AlbumTitle,
                BridgeCandidateEditField.CatalogNumber,
            },
            marked);
    }

    /// <summary>The dot names its field, so a test or an assistive technology
    /// can say which value it stands after.</summary>
    [AvaloniaFact]
    public void TheDotNamesTheFieldItStandsAfter()
    {
        var entry = ImportCandidateFixtures.FieldProvenance()
            .Single(item => item.Field == BridgeCandidateEditField.CatalogNumber);
        var dot = Assert.IsType<Ellipse>(FieldOriginDot.For(entry));
        Assert.Equal("origin-dot-catalog_number", dot.Name);
    }

    /// <summary>What stands behind the dot: every catalog describing the
    /// release and what each one states.</summary>
    [AvaloniaFact]
    public void TheHoverListsBothCatalogsReadings()
    {
        var entry = ImportCandidateFixtures.FieldProvenance()
            .Single(item => item.Field == BridgeCandidateEditField.CatalogNumber);

        var lines = FieldOriginDot.Lines(entry)
            .GetLogicalDescendants()
            .OfType<TextBlock>()
            .Select(text => text.Text)
            .ToArray();

        Assert.Contains(
            BaeBridgeMethods.BridgeCatalogName(BridgeCatalog.MusicBrainz),
            lines);
        Assert.Contains(
            BaeBridgeMethods.BridgeCatalogName(BridgeCatalog.Discogs),
            lines);
        Assert.Contains("CAT-1", lines);
        Assert.Contains("CAT-1-A", lines);
    }

    /// <summary>A value a person typed says so under the readings, so the
    /// hover explains the grey dot rather than leaving it to be guessed
    /// at.</summary>
    [AvaloniaFact]
    public void ATypedValueSaysSoInItsHover()
    {
        var entry = ImportCandidateFixtures.FieldProvenance()
            .Single(item => item.Field == BridgeCandidateEditField.AlbumTitle);

        var lines = FieldOriginDot.Lines(entry)
            .GetLogicalDescendants()
            .OfType<TextBlock>()
            .Select(text => text.Text)
            .ToArray();

        Assert.Contains(Loc.Core("core.field.origin.typed"), lines);
    }

    /// <summary>The import pane's RELEASE grid puts the dots beside the values
    /// core marked, and beside no others.</summary>
    [AvaloniaFact]
    public void TheImportGridDrawsTheDotsCoreMarked()
    {
        var section = ImportMetadataSourceSectionTests.BuildSection();
        var window = new Window { Width = 900, Height = 900, Content = section };
        window.Show();
        try
        {
            var dots = section.GetLogicalDescendants()
                .OfType<Ellipse>()
                .Select(dot => dot.Name)
                .Where(name => name?.StartsWith("origin-dot-") == true)
                .ToArray();
            Assert.Equal(
                new[] { "origin-dot-album_title", "origin-dot-catalog_number" },
                dots);
        }
        finally
        {
            window.Close();
        }
    }
}

File: bae-macos/bae/baeTests/FieldOriginDotTests.swift
import AppKit
import BaeKit
import SwiftUI
import Testing

@testable import bae

/// What the dot after a release field says, and when it says nothing.
@MainActor
@Suite("Field origin dots")
struct FieldOriginDotTests {
    /// What one field's dot draws, at a size that holds nothing else.
    private func dotPixels(
        _ provenance: BridgeFieldProvenance?
    ) async throws -> Data {
        let size = NSSize(width: 16, height: 16)
        return try await FindOnlineRendering.pixels(
            Group {
                if let provenance {
                    FieldOriginDot(provenance: provenance)
                }
            }
            .frame(width: size.width, height: size.height),
            size: size
        )
    }

    /// A field core marked draws a dot, and a field it marked nothing for
    /// draws exactly what no dot at all draws. Which two of the eight are
    /// marked is core's answer — a label the person typed, and a catalog
    /// number the catalogs state differently.
    @Test("a dot is drawn for the marked fields and for no others")
    func aDotIsDrawnForTheMarkedFieldsAndForNoOthers() async throws {
        let blank = try await dotPixels(nil)
        var marked: Set<BridgeCandidateEditField> = []
        for entry in PreviewData.fieldProvenance()
        where try await dotPixels(entry) != blank {
            marked.insert(entry.field)
        }
        #expect(marked == [.label, .catalogNumber])
    }

    /// The grid puts those dots in the form: drawing it with what core says
    /// differs from drawing it with nothing said, and a form core marked no
    /// field of draws the same as one it described not at all.
    @Test("the grid draws the dots and only when core marks one")
    func theGridDrawsTheDotsAndOnlyWhenCoreMarksOne() async throws {
        func grid(_ provenance: [BridgeFieldProvenance]) async throws -> Data {
            try await FindOnlineRendering.pixels(
                ReleasePressingFieldsGrid(
                    values: PreviewData.editMetadataDraft(trackCount: 1),
                    provenance: provenance,
                    writer: ReleaseFieldWriter { _, _ in },
                    editingCommands: EditingCommitCommands()
                )
                .padding(16)
                .environment(Library.stub())
                .environment(UiStore()),
                size: NSSize(width: 520, height: 220)
            )
        }
        let undescribed = try await grid([])
        let unmarked = try await grid(
            PreviewData.fieldProvenance()
                .map {
                    BridgeFieldProvenance(
                        field: $0.field,
                        origin: $0.origin,
                        claims: $0.claims,
                        dot: nil
                    )
                }
        )
        #expect(
            unmarked == undescribed,
            "a form core marked no field of draws no dots"
        )
        #expect(
            try await grid(PreviewData.fieldProvenance()) != undescribed,
            "the fields core marked draw theirs"
        )
    }

    /// A field a person types in is theirs before the form is ever saved, and
    /// emptying it leaves no value for an origin to describe. Core's rule,
    /// asked rather than re-derived, so the sheet's dot matches the one the
    /// stored release draws.
    @Test("typing marks the field and emptying it unmarks it")
    func typingMarksTheFieldAndEmptyingItUnmarksIt() async throws {
        let seed = PreviewData.releaseEditSeed(trackCount: 1)
        let session = ReleaseMetadataEditSession(
            releaseId: "release-1",
            seed: seed,
            save: { _, _ in },
            reset: { _ in
                BridgeReleaseFormReset(
                    edit: seed.edit,
                    fieldProvenance: seed.fieldProvenance
                )
            }
        )

        await session.fieldWriter.setField(.country, "JP")
        #expect(session.fieldProvenance.forField(.country)?.dot == .typed)
        #expect(session.form.origins.country == .typed)

        await session.fieldWriter.setField(.country, "")
        #expect(session.fieldProvenance.forField(.country)?.dot == nil)
        #expect(session.form.origins.country == nil)
    }

    /// Typing over a field the catalogs disagree about leaves the dot saying
    /// so: the readings behind it are still what it invites a look at.
    @Test("a disagreement outlives the value being typed over")
    func aDisagreementOutlivesTheValueBeingTypedOver() async throws {
        let seed = PreviewData.releaseEditSeed(trackCount: 1)
        let session = ReleaseMetadataEditSession(
            releaseId: "release-1",
            seed: seed,
            save: { _, _ in },
            reset: { _ in
                BridgeReleaseFormReset(
                    edit: seed.edit,
                    fieldProvenance: seed.fieldProvenance
                )
            }
        )

        await session.fieldWriter.setField(.catalogNumber, "CAT-0002")
        let entry = try #require(
            session.fieldProvenance.forField(.catalogNumber)
        )
        #expect(entry.origin == .typed)
        #expect(entry.dot == .disagreement)
    }

    /// The hover behind a dot names every catalog describing the release and
    /// what each one states, plus where the value in the field came from when
    /// it was not a catalog.
    @Test("the hover lists both catalogs' readings")
    func theHoverListsBothCatalogsReadings() async throws {
        let lines = try await FindOnlineRendering.text(
            FieldOriginPopover(
                provenance: PreviewData.disagreeingCatalogNumber
            ),
            size: NSSize(width: 260, height: 90)
        )
        for expected in ["MusicBrainz", "Discogs", "CAT-0001", "CAT-0001-A"] {
            #expect(
                lines.contains {
                    $0.localizedCaseInsensitiveContains(expected)
                },
                "the hover read: \(lines)"
            )
        }
    }

    /// A value a person typed says so under the readings, so the hover
    /// explains the grey dot rather than leaving it to be guessed at.
    @Test("a typed value says so in its hover")
    func aTypedValueSaysSoInItsHover() async throws {
        let lines = try await FindOnlineRendering.text(
            FieldOriginPopover(provenance: PreviewData.typedLabel),
            size: NSSize(width: 260, height: 110)
        )
        let typed = coreString("core.field.origin.typed")
        #expect(
            lines.contains { $0.localizedCaseInsensitiveContains(typed) },
            "the hover read: \(lines)"
        )
    }
}
